apps/heft/src/cli/HeftCommandLineUtilities.ts

```diff
@@ -249,60 +141,88 @@ export class HeftCommandLineUtilities {
   public registerStringListParameter(
     options: IRegisterParameterWithArgumentOptions
   ): IHeftStringListParameter {
-    const actionParameterMap: Map<CommandLineAction, CommandLineStringListParameter> = new Map();
-    for (const action of this._getActions(options.associatedActionNames)) {
+    return this._registerParameter(
+      options,
+      (action: CommandLineAction) => action.defineStringListParameter(options),
+      (parameter: CommandLineStringListParameter) => parameter.values,
+      (parameter: CommandLineStringListParameter) => (parameter.values || []).length > 0
+    );
+  }
+
+  private _registerParameter<
+    TTSCommandLineParameter extends CommandLineParameter,
+    TValue extends CustomActionParameterType
+  >(
+    options: IRegisterParameterOptions,
+    defineParameterForAction: (action: CommandLineAction) => TTSCommandLineParameter,
+    getParameterValue: (parameter: TTSCommandLineParameter) => TValue | undefined,
+    getValueIsProvided: (parameter: TTSCommandLineParameter) => boolean
+  ): IHeftBaseParameter<TValue> {
+    const actionParameterMap: Map<CommandLineAction, TTSCommandLineParameter> = new Map();
+    for (const action of this._getActions(options.associatedActionNames, options.parameterLongName)) {
       this._verifyUniqueParameterName(action, options);
-      const parameter: CommandLineStringListParameter = action.defineStringListParameter(options);
+      const parameter: TTSCommandLineParameter = defineParameterForAction(action);
       actionParameterMap.set(action, parameter);
     }
-    const parameterObject: Partial<IHeftStringListParameter> = {};
-    Object.defineProperties(parameterObject, {
-      values: {
-        get: (): readonly string[] | undefined => {
-          this._verifyParametersProcessed(options.parameterLongName);
-          if (this._commandLineParser.selectedAction) {
-            return actionParameterMap.get(this._commandLineParser.selectedAction)?.values;
+
+    const self: HeftCommandLineUtilities = this;
```

Comment:
Personally, I prefer arrow functions over `self = this`, but I understand that's a stylistic discussion. Let me know if we have a strong opinion in Rushstack / on our team. Thanks.

Reply:
Yeah I do too. Unfortunately, the `this` value in the `parameterObject` will refer to `parameterObject`. I changed it back to use `Object.defineProperties` to use lambdas instead.